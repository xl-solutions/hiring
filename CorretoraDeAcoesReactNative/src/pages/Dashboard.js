import React, {Component} from 'react';
import {
  ActivityIndicator,
  Text,
  View,
  TouchableOpacity,
  FlatList,
} from 'react-native';

import {getLastQuote, searchStock} from '../wrapper/api';
import {Card, Input} from 'react-native-elements';
import StockInformationDashboard from '../components/StockInformationDashboard';

export class Dashboard extends Component {
  state = {
    stockSelected: undefined,
    searchText: '',
    searchableStocks: [],
    loading: false,
    isFetchingQuoteInformation: false,
    quoteInformation: undefined,
  };

  componentDidMount = () => {};

  setValue = callback => {
    this.setState(state => ({
      value: callback(state.value),
    }));
  };

  handleSearchText = async () => {
    if (this.state.searchText === '') {
      this.setState({searchableStocks: []});
      return;
    }
    this.setState({loading: true});
    let searchResult = await searchStock(this.state.searchText);
    this.setState({loading: false, searchableStocks: searchResult});
  };

  getStockInformation = async stock => {
    this.setState({
      isFetchingQuoteInformation: true,
    });
    let quoteData = await getLastQuote(stock.symbol);
    console.log('quoteData', quoteData);
    this.setState({
      isFetchingQuoteInformation: false,
      quoteInformation: quoteData,
    });

    console.log(quoteData);
  };

  renderListItem = ({item}) => (
    <TouchableOpacity
      key={item.name}
      style={{margin: 5}}
      onPress={() => {
        this.setState({
          searchableStocks: [],
          stockSelected: item,
          searchText: '',
        });
        this.getStockInformation(item);
      }}>
      <View style={{flexDirection: 'row'}}>
        <Text style={{fontWeight: 'bold', marginBottom: 2}}>{item.symbol}</Text>
        <Text style={{fontStyle: 'italic', marginBottom: 2}}>
          {` - Currency: ${item.currency} `}
        </Text>
      </View>
      <Text>{`Type: ${item.type} - Region: ${item.region}`}</Text>
      <Text>{item.name}</Text>
    </TouchableOpacity>
  );

  render() {
    return (
      <View style={{marginHorizontal: 5, marginVertical: 10}}>
        <Input
          value={this.state.searchText}
          onChangeText={text => {
            this.setState({searchText: text}, () =>
              this.handleSearchText(text),
            );
          }}
          placeholder={'Search a stock'}
        />
        <View style={{position: 'absolute'}}>
          {this.state.loading ? (
            <ActivityIndicator size={16} color={'blue'} />
          ) : this.state.searchableStocks.length > 0 ? (
            <FlatList
              data={this.state.searchableStocks}
              renderItem={this.renderListItem}
              keyExtractor={item => item.id}
              style={{
                height: 200,
                flexGrow: 0,
                zIndex: 2,
              }}
            />
          ) : undefined}
        </View>
        {this.state.isFetchingQuoteInformation ? (
          <ActivityIndicator size={24} color={'blue'} />
        ) : this.state.quoteInformation !== undefined ? (
          <View style={{zIndex: 1}}>
            <View style={{backgroundColor: 'white'}}>
              <Card.Title>{this.state.quoteInformation.symbol}</Card.Title>
              <Text>Open: {this.state.quoteInformation.open}</Text>
              <Text>High: {this.state.quoteInformation.high}</Text>
              <Text>Low: {this.state.quoteInformation.low}</Text>
              <Text>Price: {this.state.quoteInformation.price}</Text>
              <Text>Volume: {this.state.quoteInformation.volume}</Text>
              <Text>
                Latest trading day:{' '}
                {this.state.quoteInformation.latestTradingDay}
              </Text>
              <Text>
                Previous close: {this.state.quoteInformation.previousClose}
              </Text>
              <Text>Change: {this.state.quoteInformation.change}</Text>
              <Text>
                Change percent: {this.state.quoteInformation.changePercent}
              </Text>
            </View>
            <StockInformationDashboard
              stock={this.state.quoteInformation.symbol}
            />
          </View>
        ) : undefined}
      </View>
    );
  }
}

export default Dashboard;
