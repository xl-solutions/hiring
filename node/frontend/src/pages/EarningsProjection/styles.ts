import styled from 'styled-components';

export const Container = styled.div`
	margin-top: 30px;

h2 {
	text-align: center;
	margin-bottom: 30px;
}
`;

export const Content = styled.div`
	margin-top: 40px;
`;

export const Row = styled.div`
	display: flex;
  flex-direction: row;
	gap: 50px;
	justify-content: center;
`;

export const Col = styled.div`
	display: flex;
	flex-direction: column;
`;
