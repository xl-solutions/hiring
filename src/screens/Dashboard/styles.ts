import styled from 'styled-components/native';
import { FlatList } from 'react-native';
import { RFValue } from 'react-native-responsive-fontsize';
// import { RFValue, RFPercentage } from 'react-native-responsive-fontsize';

type BestMatches = {
  [key in BestMatchesAttributions]: string;
};

type BestMatchesAttributions =
  | '1. symbol'
  | '2. name'
  | '3. type'
  | '4. region'
  | '5. marketOpen'
  | '6. marketClose'
  | '7. timezone'
  | '8. currency'
  | '9. matchScore';

export const Container = styled.View`
  flex: 1;

  padding: ${RFValue(10)}px ${RFValue(20)}px 0 ${RFValue(20)}px;
  background-color: ${({ theme }) => theme.screen.background};
`;

export const Title = styled.Text`
  font-family: ${({ theme }) => theme.fonts.Medium};
  font-size: ${({ theme }) => RFValue(theme.fontScale.md)}px;

  line-height: ${RFValue(21.6)}px;
  font-style: normal;
  text-align: left;

  color: ${({ theme }) => theme.neutralColors.dark['dark-light']};
`;

export const LitActionsMercado = styled(
  FlatList as new () => FlatList<BestMatches>,
).attrs({
  showsVerticalScrollIndicator: false,
  contentContainerStyle: { paddingHorizontal: 0, paddingVertical: 20 },
})`
  width: 100%;
  padding: 0px;
  background-color: ${({ theme }) => theme.screen.background};
`;
