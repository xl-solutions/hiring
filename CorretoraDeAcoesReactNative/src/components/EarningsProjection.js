import React, {useState} from 'react';
import {
  View,
  Text,
  Modal,
  ToastAndroid,
  TouchableOpacity,
  ActivityIndicator,
} from 'react-native';
import {Input, Button, Card, Divider} from 'react-native-elements';
import {getDailyAdjustedSeries} from '../wrapper/api';
import DateTimePickerModal from 'react-native-modal-datetime-picker';
import moment from 'moment';
import styles from './styles';

const EarningsProjection = props => {
  const [marketShareAmount, setMarketShareAmount] = useState('1');
  const [pastDate, setPastDate] = useState();
  const [datePickerVisible, setDatePickerVisible] = useState(false);
  const [loading, setLoading] = useState(false);
  const [projection, setProjection] = useState(undefined);

  const calculateProjection = async () => {
    let data = await getDailyAdjustedSeries(props.symbol);
    let selectedDate = data.find(element => element.date === pastDate);
    let lastCoted = data[0];
    let valueForPastDate =
      parseFloat(selectedDate.close) * parseInt(marketShareAmount, 10);
    let valueForCurrentDate =
      parseFloat(lastCoted.close) * parseInt(marketShareAmount, 10);
    setProjection({
      symbol: props.symbol,
      boughtDate: pastDate,
      valueForGivenDate: selectedDate.close,
      valueForCurrentDate: lastCoted.close,
      lastRefreshDate: lastCoted.date,
      profit: valueForCurrentDate - valueForPastDate,
    });
    setLoading(false);
  };

  const hideDatePicker = () => setDatePickerVisible(false);

  const handleConfirm = date => {
    let fomartedDate = moment(date).format('YYYY-MM-DD');
    if (moment(date).isoWeekday() !== 6 && moment(date).isoWeekday() !== 7) {
      setPastDate(fomartedDate);
    } else {
      ToastAndroid.show('Weekends are not permitted', ToastAndroid.SHORT);
    }
    hideDatePicker();
  };

  return (
    <Modal visible={props.visible} nimationType="fade" transparent={true}>
      <View style={styles.modalOverlay}>
        <View style={styles.modalContainer}>
          <DateTimePickerModal
            isVisible={datePickerVisible}
            mode="date"
            maximumDate={new Date()}
            onConfirm={handleConfirm}
            onCancel={hideDatePicker}
          />
          <Text style={styles.titleModal}>Earnings projection</Text>
          <Divider style={styles.marginVertical5} />
          <View style={styles.directionRow}>
            <TouchableOpacity
              style={styles.defaultFlex}
              onPress={() => {
                setDatePickerVisible(true);
              }}>
              <Input
                editable={false}
                placeholder="Past date"
                value={pastDate}
              />
            </TouchableOpacity>
            <View style={styles.defaultFlex}>
              <Input
                onChangeText={text => setMarketShareAmount(text)}
                placeholder="Number of market shares"
                value={marketShareAmount}
              />
            </View>
          </View>
          <Button
            title="Calculate"
            onPress={() => {
              if (
                pastDate === undefined ||
                isNaN(parseInt(marketShareAmount, 10))
              ) {
                ToastAndroid.show(
                  'Please put valid input data',
                  ToastAndroid.SHORT,
                );
              } else {
                setLoading(true);
                calculateProjection();
              }
            }}
          />
          <View style={styles.defaultFlex}>
            {loading ? (
              <ActivityIndicator size={20} color="blue" />
            ) : projection ? (
              <Card>
                <Text>Symbol: {projection.symbol}</Text>
                <Text>Bought date: {projection.boughtDate}</Text>
                <Text>
                  Value for given date: ${projection.valueForGivenDate}
                </Text>
                <Text>
                  Value for current date: ${projection.valueForCurrentDate}
                </Text>
                <Text>Last refresh: {projection.lastRefreshDate} </Text>
                <Text>
                  Profit:{' '}
                  <Text
                    // eslint-disable-next-line react-native/no-inline-styles
                    style={{color: projection.profit < 0 ? 'red' : 'green'}}>
                    ${projection.profit.toFixed(2)}
                  </Text>
                </Text>
              </Card>
            ) : undefined}
          </View>
          <Button title="Close" onPress={() => props.close()} />
        </View>
      </View>
    </Modal>
  );
};

export default EarningsProjection;
