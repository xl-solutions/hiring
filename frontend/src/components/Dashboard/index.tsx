
import { Stocks } from '../Stocks'
import { Container } from './styles'

export function DashBoard(){
    return(
        <Container>
            <Stocks/>
        </Container>
    )
}