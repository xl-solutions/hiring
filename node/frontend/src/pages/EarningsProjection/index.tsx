import { FormEvent, useState } from 'react';
import DatePicker from 'react-datepicker';
import { TableColumn } from 'react-data-table-component';
import moment from 'moment';

import { DataTableComponent } from '../../components/DataTable';
import { Button } from '../../components/Button';

import { api } from '../../api';

import { Col, Container, Content, Row } from './styles';
import { formatValueToCurrency } from '../../utils';

interface IDataRow {
	name: string;
	purchasedAmount: number;
	purchasedAt: string;
	priceAtDate: number;
	lastPrice: number;
	capitalGains: number;
}

const columns: TableColumn<IDataRow>[] = [
	{
		name: 'Nome',
		selector: row => row.name
	},
	{
		name: 'Valor da compra',
		selector: row => formatValueToCurrency(row.purchasedAmount)
	},
	{
		name: 'Data da compra',
		selector: row => moment(row.purchasedAt).format('DD/MM/YYYY')
	},
	{
		name: 'Valor na data da compra',
		selector: row => formatValueToCurrency(row.priceAtDate)
	},
	{
		name: 'Último Fechamento',
		selector: row => formatValueToCurrency(row.lastPrice)
	},
	{
		name: 'Lucro ou perda',
		selector: row =>  formatValueToCurrency(row.capitalGains)
	}
];

export function EarningsProjection() {
	const [earningsProjection, setEarningsProjection] = useState<IDataRow[]>([]);
	const [symbol, setSymbol] = useState('');
	const [purchasedAmount, setPurchasedAmount] = useState(0);
	const [purchasedDate, setPurchasedDate] = useState(new Date());

	async function handleSubmit(event: FormEvent) {
		event.preventDefault();

		try {
			const { data } = await api.get(
				`${process.env.REACT_APP_API_URL}/${symbol}/gains`,
				{ params: {
					purchasedAmount,
					purchasedAt: moment(purchasedDate).format('YYYY-MM-DD')
				} }
			);

			const dataEarningsProjection: IDataRow[]  = [];
			dataEarningsProjection.push(data);
			setEarningsProjection(dataEarningsProjection);

		} catch (error) {
			console.error(error.response.data);
		}

	}

	return (
		<Container>
			<h2>Projeção de ganhos</h2>
			<form onSubmit={handleSubmit}>
				<Row>
					<Col>
						<label>Simbolo da ação:</label>
						<input
							type="text"
							placeholder='Ex. PETR4.SA'
							onChange={(e) => setSymbol(e.target.value)}
						/>
					</Col>
					<Col>
						<label>Valor comprado R$: </label>
						<input
							type="number"
							placeholder='Ex. R$ 250'
							onChange={(e) => setPurchasedAmount(Number(e.target.value))}

						/>
					</Col>
					<Col>
						<label>Data da compra:</label>
						<DatePicker
							selected={purchasedDate}
							onChange={(date: Date) => setPurchasedDate(date)}
							isClearable
							showYearDropdown
							dateFormat="dd/MM/yyyy"
							filterDate = {(date) => {
								return new Date() > date;
							}}
						/>
					</Col>
					<Col>
						<Button type="submit">
							Projetar ganhos
						</Button>
					</Col>
				</Row>
			</form>
			<Content>
				<DataTableComponent
					columns={columns}
					data={earningsProjection}
				/>
			</Content>
		</Container>
	);
}
