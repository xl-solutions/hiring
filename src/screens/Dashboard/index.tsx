/* eslint-disable react-native/no-inline-styles */
import React, { useState } from 'react';
import {
  KeyboardAvoidingView,
  TouchableWithoutFeedback,
  Keyboard,
  Platform,
  View,
} from 'react-native';
import { useTheme } from 'styled-components/native';
import { ActionItem } from '../../components/ActionItem';
import { HeaderBase } from '../../components/HeaderBase';
import { InputFormController } from '../../components/InputFormController';
import { LoadButton } from '../../components/Loading/LoadButton';
import { StatusBarBase } from '../../components/StatusBarBase';
import { useFetch } from '../../hooks/fetchData';

import { Container, LitActionsMercado, Title } from './styles';

let timeOuId: any;

function Dashboard() {
  const { search, bestMatchesActions, loading } = useFetch();
  const { statusBar, neutralColors } = useTheme();

  const [isFocused, setIsFocused] = useState<boolean>(false);
  const [searchText, setSearchText] = useState<string>();

  const debounce = (func: Function, delay: number) => {
    return (...args: any[]) => {
      if (timeOuId) {
        clearTimeout(timeOuId);
      }

      timeOuId = setTimeout(() => func.apply(null, args), delay);
    };
  };

  async function fetchSearch(value: string) {
    await search(value);
  }

  const debounceSearch = debounce(fetchSearch, 600);

  const handleOnChange = (text: string) => {
    setSearchText(text);
    debounceSearch(searchText);
  };

  function listHeaderComponent() {
    return (
      <View style={{ flexDirection: 'row', marginBottom: 20 }}>
        <Title>Ações encontradas</Title>
        {loading ? (
          <LoadButton color={neutralColors.dark['dark-light']} size={20} />
        ) : (
          <Title style={{ marginLeft: 20 }}>
            {' '}
            {bestMatchesActions?.length}
          </Title>
        )}
      </View>
    );
  }

  return (
    <>
      <KeyboardAvoidingView
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
        style={{ flex: 1 }}>
        <StatusBarBase
          barStyle="dark-content"
          backgroundColor={statusBar.backgroundDefault}
          animated
          translucent={false}
        />
        <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
          <Container>
            <HeaderBase title="Google Finances" icon={false} />

            <InputFormController
              testID="test-input-search"
              placeholder="Pesquisar ações, ETFs e outros para comprar"
              name="search"
              isFocused={isFocused}
              onFocus={() => setIsFocused(true)}
              onBlur={() => setIsFocused(false)}
              iconName="search"
              onChangeText={event => handleOnChange(event)}

              // onChange={event => handleOnChange(event.nativeEvent.text)}
            />

            <LitActionsMercado
              testID="test-LitActionsMercado"
              data={bestMatchesActions}
              ListHeaderComponent={listHeaderComponent}
              keyExtractor={item => String(item['1. symbol'])}
              renderItem={({ item }) => <ActionItem item={item} />}
            />
          </Container>
        </TouchableWithoutFeedback>
      </KeyboardAvoidingView>
    </>
  );
}

export { Dashboard };
