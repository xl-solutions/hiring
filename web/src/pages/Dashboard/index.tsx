import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Header } from '../../components/Header';
import { Container, Content } from './styles';
import api from '../../services/api';
import { GeneralStockInfo } from '../../components/GeneralStockInfo';
import { ProjectionGains } from '../../components/ProjectionGains';
import { CompareStock } from '../../components/CompareStock';
import { updateCurrentStock } from '../../store/modules/currentStock/actions';
import { IState } from '../../store';
import { IStock } from '../../store/modules/currentStock/types';

export interface CurrentStockProps {
  name: string;
  lastPrice?: number;
  company?: string;
  region?: string;
  currency?: string;
}

export default function Dashboard() {
  const dispatch = useDispatch();
  const currentStock = useSelector<IState, IStock>(
    (state) => state.currentStock.stock
  );

  async function searchQuoteFromStock(
    stock_name: string,
    stock_company?: string,
    stock_region?: string,
    stock_currency?: string
  ) {
    const { data } = await api.get(`/${stock_name}/quote`);
    const stock = {
      name: stock_name,
      lastPrice: data.lastPrice,
      company: stock_company || currentStock.company,
      region: stock_region || currentStock.region,
      currency: stock_currency || currentStock.currency,
    };

    dispatch(updateCurrentStock(stock));
  }

  useEffect(() => {
    searchQuoteFromStock(currentStock.name);
  }, []);

  return (
    <Container>
      <Header withSearchBar fetchData={searchQuoteFromStock} />
      <Content>
        <GeneralStockInfo />
        <h2>Projeção de ganhos por compra</h2>
        <ProjectionGains />
        <h2>Comparar com outros ativos</h2>
        <CompareStock />
      </Content>
    </Container>
  );
}
