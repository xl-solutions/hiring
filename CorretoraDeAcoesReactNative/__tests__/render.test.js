import React from 'react';
import renderer from 'react-test-renderer';
import EarningsProjection from '../src/components/EarningsProjection';
import Dashboard from '../src/pages/Dashboard';

it('renders correctly', () => {
  const tree = renderer.create(<EarningsProjection />).toJSON();
  expect(tree).toMatchSnapshot();
});
