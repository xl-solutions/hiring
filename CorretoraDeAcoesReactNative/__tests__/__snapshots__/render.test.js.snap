// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "marginHorizontal": 5,
      "marginVertical": 10,
    }
  }
>
  <View
    style={
      Object {
        "paddingHorizontal": 10,
        "width": "100%",
      }
    }
  >
    <View
      collapsable={false}
      nativeID="animatedComponent"
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <TextInput
        editable={true}
        onChangeText={[Function]}
        placeholder="Search a stock"
        placeholderTextColor="#86939e"
        style={
          Object {
            "color": "#242424",
            "flex": 1,
            "fontSize": 18,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <Text
      style={
        Object {
          "color": "#ff190c",
          "fontSize": 12,
          "margin": 5,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "left": 20,
        "position": "absolute",
        "top": 50,
        "width": "90%",
        "zIndex": 2,
      }
    }
  />
</View>
`;
